File: Java/152. Maximum Product Subarray.java
class Solution {
    public int maxProduct(int[] nums) {
        int maxsofar = nums[0];
        int minsofar = maxsofar;
        int result = maxsofar;
        
        for(int i=1; i<nums.length; i++){
            int temp_max = Math.max(nums[i], Math.max(maxsofar*nums[i], minsofar*nums[i]));
            minsofar = Math.min(nums[i], Math.min(maxsofar*nums[i], minsofar*nums[i]));
            maxsofar = temp_max;
            
            result = Math.max(result, maxsofar);
        }
        
        return result;
    }
}

File: Python3/152. Maximum Product Subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        
        maxsofar = nums[0]
        minsofar = nums[0]
        result = maxsofar
        
        for num in nums[1:]:
            # use a temp var to store the max value so far
            # it will be used for computing minsofar
            temp_max = max(num, num*maxsofar, num*minsofar)
            minsofar = min(num, num*maxsofar, num*minsofar)
            # update maxsofar
            maxsofar = temp_max
            
            # update result if possible
            result = max(result, maxsofar)
            
        return result